File: mock-api/api/routes/compare.py
"""Compare routes — /api/v1/projects/:id/compare/*"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from typing import Optional

from ..db.session import get_db
from ..db.in_memory import InMemoryDatabase
from ..middleware.auth import get_current_user, require_project_member
from ..models.domain import User
from ..services.errors import not_found
from ..schemas import (
    CompareResponse, CompareDocumentsResponse, CompareDocumentDetailResponse,
)

router = APIRouter(tags=["compare"])


@router.get("/projects/{project_id}/compare",
            responses={200: {"model": CompareResponse}})
def compare(
    project_id: str,
    current: str = Query(...),
    baseline: str = Query(...),
    current_user: User = Depends(get_current_user),
    db: InMemoryDatabase = Depends(get_db),
):
    if not db.projects.get(project_id):
        raise not_found("Project", project_id)
    require_project_member(project_id, current_user, db)
    cr = db.compare.get_or_create(project_id, current, baseline)
    diffs = db.compare.list_diffs(cr.id)
    # Resolve version labels
    def _resolve_ref(ref: str):
        for v in db.versions.list_for_project(project_id):
            if v.commit_sha == ref or v.tag == ref or v.id == ref:
                return {"ref": ref, "version": v.tag, "branch": v.branch}
        return {"ref": ref, "version": None, "branch": "main"}

    changed_docs = []
    for d in diffs:
        doc = db.documents.get(d.document_id)
        if doc:
            changed_docs.append({
                "document_id": d.document_id,
                "name": doc.name,
                "process": doc.process,
                "diff_type": d.diff_type,
                "sections_changed": d.sections_changed,
            })

    return {
        "current": _resolve_ref(current),
        "baseline": _resolve_ref(baseline),
        "summary": cr.diff_summary,
        "changed_documents": changed_docs,
    }


@router.get("/projects/{project_id}/compare/documents",
            responses={200: {"model": CompareDocumentsResponse}})
def compare_documents(
    project_id: str,
    current: str = Query(...),
    baseline: str = Query(...),
    current_user: User = Depends(get_current_user),
    db: InMemoryDatabase = Depends(get_db),
):
    if not db.projects.get(project_id):
        raise not_found("Project", project_id)
    require_project_member(project_id, current_user, db)
    cr = db.compare.get_or_create(project_id, current, baseline)
    diffs = db.compare.list_diffs(cr.id)
    result = []
    for d in diffs:
        doc = db.documents.get(d.document_id)
        if doc:
            result.append({
                "document_id": d.document_id,
                "name": doc.name,
                "process": doc.process,
                "diff_type": d.diff_type,
                "sections_changed": d.sections_changed,
            })
    return {"documents": result, "summary": cr.diff_summary}


@router.get("/projects/{project_id}/compare/documents/{doc_id}",
            responses={200: {"model": CompareDocumentDetailResponse}})
def compare_document_detail(
    project_id: str,
    doc_id: str,
    current: str = Query(...),
    baseline: str = Query(...),
    current_user: User = Depends(get_current_user),
    db: InMemoryDatabase = Depends(get_db),
):
    if not db.projects.get(project_id):
        raise not_found("Project", project_id)
    require_project_member(project_id, current_user, db)
    doc = db.documents.get(doc_id)
    if not doc or doc.project_id != project_id:
        raise not_found("Document", doc_id)
    cr = db.compare.get_or_create(project_id, current, baseline)
    diff = db.compare.get_document_diff(cr.id, doc_id)
    sections = db.documents.list_sections(doc_id)
    changed_keys = set(diff.sections_changed) if diff else set()
    section_diffs = []
    for s in sections:
        dt = "changed" if s.section_key in changed_keys else "unchanged"
        section_diffs.append({
            "key": s.section_key,
            "title": s.title,
            "diff_type": dt,
            "current_content": s.content,
            "baseline_content": s.content if dt == "unchanged" else "[previous version content]",
        })
    return {"document_name": doc.name, "sections": section_diffs}

File: web-app/src/services/api/compare.ts
import { http } from '../../lib/http'
import type { CompareResult, CompareDocumentDiff, CompareChangedDoc } from '../../types'
import { mapCompare, mapCompareDocumentDiff } from '../mappers'

export const compareApi = {
  summary: async (projectId: string, current: string, baseline: string): Promise<CompareResult> => {
    const r = await http.get<Parameters<typeof mapCompare>[0]>(`/projects/${projectId}/compare`, {
      current,
      baseline,
    })
    return mapCompare(r)
  },
  documents: async (
    projectId: string,
    current: string,
    baseline: string,
  ): Promise<{ documents: CompareChangedDoc[]; summary: CompareResult['summary'] }> => {
    const r = await http.get<{
      documents: { document_id: string; name: string; process: string; diff_type: string; sections_changed: string[] }[]
      summary: CompareResult['summary']
    }>(`/projects/${projectId}/compare/documents`, { current, baseline })
    return {
      documents: r.documents.map((d) => ({
        documentId: d.document_id,
        name: d.name,
        process: d.process,
        diffType: d.diff_type as CompareChangedDoc['diffType'],
        sectionsChanged: d.sections_changed,
      })),
      summary: r.summary,
    }
  },
  documentDetail: async (
    projectId: string,
    docId: string,
    current: string,
    baseline: string,
  ): Promise<CompareDocumentDiff> => {
    const r = await http.get<Parameters<typeof mapCompareDocumentDiff>[0]>(
      `/projects/${projectId}/compare/documents/${docId}`,
      { current, baseline },
    )
    return mapCompareDocumentDiff(r)
  },
}
